plugins/catalog-backend/src/ingestion/processors/AnnotateLocationEntityProcessor.ts

```diff
@@ -15,31 +15,79 @@
  */
 
 import {
+  EDIT_URL_ANNOTATION,
   Entity,
-  LOCATION_ANNOTATION,
   LocationSpec,
+  LOCATION_ANNOTATION,
   ORIGIN_LOCATION_ANNOTATION,
+  SOURCE_LOCATION_ANNOTATION,
+  VIEW_URL_ANNOTATION,
 } from '@backstage/catalog-model';
+import { ScmIntegrationRegistry } from '@backstage/integration';
 import lodash from 'lodash';
 import { CatalogProcessor, CatalogProcessorEmit } from './types';
+import gitUrlParse from 'git-url-parse';
+
+type Options = {
+  integrations: ScmIntegrationRegistry;
+};
 
 export class AnnotateLocationEntityProcessor implements CatalogProcessor {
+  constructor(private readonly options: Options) {}
+
   async preProcessEntity(
     entity: Entity,
     location: LocationSpec,
     _: CatalogProcessorEmit,
     originLocation: LocationSpec,
   ): Promise<Entity> {
-    return lodash.merge(
-      {
-        metadata: {
-          annotations: {
-            [LOCATION_ANNOTATION]: `${location.type}:${location.target}`,
-            [ORIGIN_LOCATION_ANNOTATION]: `${originLocation.type}:${originLocation.target}`,
-          },
-        },
-      },
-      entity,
-    );
+    const { integrations } = this.options;
+
+    const annotations: Record<string, string> = {
+      [LOCATION_ANNOTATION]: `${location.type}:${location.target}`,
+      [ORIGIN_LOCATION_ANNOTATION]: `${originLocation.type}:${originLocation.target}`,
+    };
+
+    if (isValidUrl(location.target)) {
+      const scmIntegration =
+        location.type === 'url'
+          ? integrations.byUrl(location.target)
+          : undefined;
+
+      annotations[VIEW_URL_ANNOTATION] = location.target;
+      const editUrl = scmIntegration?.resolveEditUrl(location.target);
+
+      if (editUrl) {
+        annotations[EDIT_URL_ANNOTATION] = editUrl;
+      }
+
+      annotations[SOURCE_LOCATION_ANNOTATION] = extractRepositoryRoot(
```

Comment:
Thinking about this, I don't think it's right after all. You probably want to do something like "if the last path segment matches `/\.ya?ml$/`, then strip it off". Because in a monorepo situation where the file is not at the root of the repo but rather together with the module that it describes, you want to go the folder that the file is in.

Does `scmIntegration.resolveUrl({ base: location.target, url: './' })` or something like that work for that? Can't check right now :)

Reply:
Oh right, the thing with `/tree/` vs `/blob/`. Yeah well, github at least seems to redirect properly both ways, so clicking it will actually work in practice

Reply:
Yes in general `./` would be better than `/`, especially in Mono Repositories.

I will play around a bit with how that would work with GitLab and Bitbucket.